src/bootstrap/test.rs

```diff
@@ -975,9 +975,19 @@ impl Step for Compiletest {
             builder.ensure(compile::Rustc { compiler, target });
         }
 
-        if builder.no_std(target) != Some(true) {
+        if builder.no_std(target) == Some(true) {
+            // the `test` doesn't compile for no-std targets
+            builder.ensure(compile::Std { compiler, target });
+        } else {
             builder.ensure(compile::Test { compiler, target });
         }
+
+        if builder.no_std(target) == Some(true) {
```

Comment:
Should this block be combined with the block above (line 980)?

Reply:
At first, I've written it in the form the second `if` is merged with the first `if`.

Later, I divided them into two `if`s with clear intention.  They are representing the two different purposes/concerns.

Let me explain the intention(s) below.

## The first `if`

```
        if builder.no_std(target) == Some(true) {
            // the `test` doesn't compile for no-std targets
            builder.ensure(compile::Std { compiler, target });
        } else {
            builder.ensure(compile::Test { compiler, target });
        }
```
The first `if` serves for building `compile::Test`. Unfortunately, you can't build `compile::Test` for `no_std`.  In that case, build `compile::Std` instead.  It should be in the form of `A` or `B`.  Adding unrelated `C` in one arm is not a good idea.

It's also a verbatim copy of @japaric's code found in `dist.rs` and supposed to be like a *idiom* which would be used in various places consistently.  In the current `bootstrap` code, `test` support for `no_std` target is quite minimum.  To widen the supported coverage for `no_std`, we might want to grep `builder.ensure(compile::Test` and replace it with this snippet.

But in future, `no_std` might get `libtest` support.  At that time the whole `if` must be reverted back to the original single line:
```
        builder.ensure(compile::Test { compiler, target });
```
If you merged two `if`s, the whole change would be much less trivial.

## The second `if`

```
        if builder.no_std(target) == Some(true) {
            // for no_std run-make (e.g. thumb*),
            // we need a host compiler which is called by cargo.
            builder.ensure(compile::Std { compiler, target: compiler.host });
        }
```
The second `if` serves for a totally different purpose. For the majority of `run-make` tests, host compiler is NEVER needed.  It is needed because we must run `cargo` and some dependent crate (e.g. `cc`) requires host compiler to be successfully built.